File: solutions/maximum-depth-of-n-ary-tree/maximum-depth-of-n-ary-tree.cpp
// @Title: Maximum Depth of N-ary Tree
// @Author: realzhangm
// @Date: 2021-11-02T15:33:25+08:00
// @URL: https://leetcode-cn.com/problems/maximum-depth-of-n-ary-tree


/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    int maxDepth(Node* root) {
        if (root == nullptr) return 0;
        int max = 0;
        for (auto &node : root->children) {
            int node_depth = maxDepth(node);
            if (node_depth > max) {
                max = node_depth;
            }
        }
        return max + 1;
    }
};

File: solutions/maximum-depth-of-n-ary-tree/maximum-depth-of-n-ary-tree.go
// @Title: Maximum Depth of N-ary Tree
// @Author: realzhangm
// @Date: 2021-11-02T15:46:41+08:00
// @URL: https://leetcode-cn.com/problems/maximum-depth-of-n-ary-tree


/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

func maxDepth(root *Node) int {
    q := []*Node{}
    if root == nil {
        return 0
    }
    depth := 0
    q = append(q, root)

    for {
        levelCnt := len(q)
        if levelCnt == 0 {
            break
        }

        for i := 0; i < levelCnt; i++ {
            for _, childNode := range q[i].Children {
                if childNode != nil {
                    q = append(q, childNode)
                }
            }
        }

        q = q[levelCnt:len(q)]
        depth++
    }
    return depth
}
